src/liquidity_mining/invariants.rs

```diff
@@ -0,0 +1,163 @@
+use crate::{assert_eq_approx, types::Balance};
+use proptest::prelude::*;
+use sp_arithmetic::traits::One;
+use sp_arithmetic::{
+    traits::{CheckedAdd, CheckedDiv, CheckedMul, CheckedSub},
+    FixedPointNumber, FixedU128,
+};
+
+fn periods() -> impl Strategy<Value = u32> {
+    0..=u32::MAX
+}
+
+fn initial_reward_percentage() -> impl Strategy<Value = u128> {
+    1..=999_999_999_999_999_999_u128 //< FixedU128::one()
+}
+
+fn scale_coef() -> impl Strategy<Value = u32> {
+    0..=(u32::MAX / 2) //half of periods() so we are able to test `periods > scale_coef` case
+}
+
+proptest! {
+    #![proptest_config(ProptestConfig::with_cases(1_000))]
+    #[test]
+    fn calculate_loyalty_multiplier(
+        periods in periods(),
+        initial_reward_percentage in initial_reward_percentage(),
+        scale_coef in scale_coef(),
+    ) {
+        let b = FixedU128::from_inner(initial_reward_percentage);
+        let multiplier = crate::liquidity_mining::calculate_loyalty_multiplier(
+            periods,
+            b,
+            scale_coef,
+        ).unwrap();
+
+        assert!(multiplier < FixedU128::one(), "loyalty multiplier is >= 1");
+
+        //b + (1 - b)/2;
+        let bound = FixedU128::one()
+                .checked_sub(&b).unwrap()
+                .checked_div(&FixedU128::from(2_u128)).unwrap()
+                .checked_add(&b).unwrap();
+
+        if periods < scale_coef {
```

Comment:
It's probably better to split into a test function where you guarantee periods <= scale_coef and a second test function where periods >= scale_coef